libkbfs/block_retrieval_queue.go

```diff
@@ -216,161 +211,37 @@ func (brq *blockRetrievalQueue) notifyWorker(priority int) {
-	ctx = CtxWithRandomIDReplayable(ctx, "prefetchForBlockID", ptr.ID.String(),
-		brq.log)
-	defer cancel()
-	prefetcher := brq.Prefetcher()
-	childPrefetchDoneCh, childPrefetchCancelCh, numBlocks :=
-		prefetcher.PrefetchAfterBlockRetrieved(block, ptr, kmd)
-
-	// If we have child blocks to prefetch, wait for them.
-	if numBlocks > 0 {
-		for i := 0; i < numBlocks; i++ {
-			select {
-			case <-childPrefetchDoneCh:
-				// We expect to receive from this channel `numBlocks` times,
-				// after which we know the subtree of this block is done
-				// prefetching.
-				continue
-			case <-ctx.Done():
-				brq.log.Warning("Prefetch canceled for block %s", ptr.ID)
-				deepPrefetchCancelCh <- struct{}{}
-				return
-			case <-childPrefetchCancelCh:
-				// One error means this block didn't finish prefetching.
-				brq.log.Warning("Prefetch canceled for block %s due to "+
-					"downstream failure", ptr.ID)
-				deepPrefetchCancelCh <- struct{}{}
-				return
-			case <-prefetcher.ShutdownCh():
-				deepPrefetchCancelCh <- struct{}{}
-				return
-			}
-		}
-	}
-
-	// Make sure we've already updated the metadata once to avoid a race.
-	<-didUpdateCh
-	// Prefetches are done. Update the caches.
-	err := brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(),
-		block, lifetime, FinishedPrefetch)
-	if err != nil {
-		brq.log.CWarningf(ctx, "Error updating cache after prefetch: %+v",
-			err)
-	}
-	dbc := brq.config.DiskBlockCache()
-	if dbc != nil {
-		err := dbc.UpdateMetadata(ctx, ptr.ID, FinishedPrefetch)
-		if err != nil {
-			brq.log.CWarningf(ctx, "Error updating disk cache after "+
-				"prefetch: %+v", err)
-			deepPrefetchCancelCh <- struct{}{}
-			return
-		}
-	}
-	brq.log.CDebugf(ctx, "Finished prefetching for block %s", ptr.ID)
-	// Now prefetching is actually done.
-	deepPrefetchDoneCh <- struct{}{}
-}
-
-// CacheAndPrefetch implements the BlockRetrieval interface for
-// blockRetrievalQueue. It also updates the LRU time for the block in the disk
-// cache.
-// `deepPrefetchDoneCh` and `deepPrefetchCancelCh` can be nil so the caller
-// doesn't always have to instantiate a channel if it doesn't care about
-// waiting for the prefetch to complete. In this case, the
-// `blockRetrievalQueue` instantiates each channel to monitor the prefetches.
-func (brq *blockRetrievalQueue) CacheAndPrefetch(ctx context.Context,
-	ptr BlockPointer, block Block, kmd KeyMetadata, priority int,
-	lifetime BlockCacheLifetime, prefetchStatus PrefetchStatus,
-	deepPrefetchDoneCh, deepPrefetchCancelCh chan<- struct{}) (err error) {
-	// Avoid having to check for nil channels below this point.
-	if deepPrefetchDoneCh == nil {
-		deepPrefetchDoneCh = make(chan struct{}, 1)
-	}
-	if deepPrefetchCancelCh == nil {
-		deepPrefetchCancelCh = make(chan struct{}, 1)
+// PutInCaches implements the BlockRetriever interface for
+// BlockRetrievalQueue.
+func (brq *blockRetrievalQueue) PutInCaches(ctx context.Context,
```

Comment:
Seems like this could be in `block_util` or something, no need to have it be in `blockRetrievalQueue` as far as I can tell.  Might be a little cleaner that way.